### PR TITLE
Mod : Fixed 2 bugs in bannerbalancing

### PR DESCRIPTION
**Pull request by namidaka**
Fixed isSwapValid for clangroups not taking in account clan group penalty

Fixed IsWeightRatioAcceptable not taking in account clan group penalty

### CODE REVIEW

File: src/Module.Server/Balancing/WeightHelpers.cs [hunk 1]
```diff
@@ -19,4 +19,9 @@ internal static float ComputeTeamWeight(List<WeightedCrpgUser> team, float p = M
         var clangroupTeam = MatchBalancingHelpers.SplitUsersIntoClanGroups(team);
         return clangroupTeam.Sum(c => c.Weight());
     }
+    internal static float ComputeTeamAbsWeight(List<WeightedCrpgUser> team, float p = MatchBalancer.PowerParameter)
```
Comment (verdie-g) [inline src/Module.Server/Balancing/WeightHelpers.cs:22]: You should probably have a warning here.

File: src/Module.Server/Balancing/WeightHelpers.cs [hunk 2]
```diff
@@ -19,4 +19,9 @@ internal static float ComputeTeamWeight(List<WeightedCrpgUser> team, float p = M
...
+    {
```
Comment (verdie-g) [inline src/Module.Server/Balancing/WeightHelpers.cs:23]: Why do we need to normal and absolute values?

Comment (verdie-g) [reply src/Module.Server/Balancing/WeightHelpers.cs:23]: Can the weight be negative?

File: src/Module.Server/Balancing/MatchBalancer.cs [hunk 3]
```diff
@@ -415,6 +416,23 @@ private bool FindAndSwapUsers(GameMatch gameMatch, float sizeScaler)
...
+            ? strongTeam.Count + 2 * sourceGroupSize - 2f * destinationGroupSize - weakTeam.Count
+            : strongTeam.Count - 2 * sourceGroupSize + 2f * destinationGroupSize - weakTeam.Count;
+
+        Vector2 oldDifferenceVector = new((strongTeam.Count - weakTeam.Count) * sizeScaler,
+            strongTeam.Sum(c => c.Weight()) - weakTeam.Sum(c => c.Weight()));
+        Vector2 newDifferenceVector = new(newTeamSizeDiff * sizeScaler, newTeamWeightDiff);
+        return newDifferenceVector.Length() < oldDifferenceVector.Length();
+    }
+
     private bool IsSwapValid(List<WeightedCrpgUser> strongTeam, List<WeightedCrpgUser> weakTeam, bool swappingFromWeakTeam,
```
Comment (verdie-g) [inline src/Module.Server/Balancing/MatchBalancer.cs:436]: Is this method still used?

**Review by verdie-g — CHANGES_REQUESTED** [review]

File: src/Module.Server/Balancing/MatchBalancer.cs (hunk 3 above)
Comment (namidaka) [reply src/Module.Server/Balancing/MatchBalancer.cs:436]: needed by a method that is bugged and need a fix ( FindAndSwapUsers )

File: src/Module.Server/Balancing/WeightHelpers.cs (hunk 2 above)
Comment (namidaka) [reply src/Module.Server/Balancing/WeightHelpers.cs:23]: Depends on weight implem. This allows us to have negative weight without breaking the algorithm

File: src/Module.Server/Balancing/WeightHelpers.cs (hunk 1 above)
Comment (namidaka) [reply src/Module.Server/Balancing/WeightHelpers.cs:22]: what warning?

**Review by verdie-g — APPROVED** [review]

**Pull request merged**